internal/webhook/sentinel/handler.go

```diff
@@ -0,0 +1,187 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package sentinel
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"slices"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+
+	"github.com/go-logr/logr"
+	admissionv1 "k8s.io/api/admission/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	corev1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/scale/scheme/autoscalingv1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+var allowedOperations = []admissionv1.Operation{admissionv1.Create, admissionv1.Connect}
+
+// Handler is the Sentinel Webhook admission handler.
+type Handler struct {
+	client.Client
+	config  *Config
+	decoder *admission.Decoder
+	logger  logr.Logger
+}
+
+// NewHandler creates a new handler for Sentinel Webhook.
+func NewHandler(mgr manager.Manager, config *Config) (*Handler, error) {
+	decoder := admission.NewDecoder(mgr.GetScheme())
+	return &Handler{
+		Client:  mgr.GetClient(),
+		config:  config,
+		decoder: decoder,
+		logger:  mgr.GetLogger().WithName(handlerName),
+	}, nil
+}
+
+// Handle handles admission requests and prevents unintended changes to resources created by etcd-druid.
+func (h *Handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	requestGKString := fmt.Sprintf("%s/%s", req.Kind.Group, req.Kind.Kind)
+	log := h.logger.WithValues("name", req.Name, "namespace", req.Namespace, "resourceGroupKind", requestGKString, "operation", req.Operation, "user", req.UserInfo.Username)
+	log.Info("Sentinel webhook invoked")
+
+	if slices.Contains(allowedOperations, req.Operation) {
+		return admission.Allowed(fmt.Sprintf("operation %s is allowed", req.Operation))
+	}
+
+	requestGK := schema.GroupKind{Group: req.Kind.Group, Kind: req.Kind.Kind}
+
+	// Leases (member and snapshot) will be periodically updated by etcd members. Allow updates to leases.
+	if requestGK == coordinationv1.SchemeGroupVersion.WithKind("Lease").GroupKind() &&
```

Comment:
Could we add a check to ensure that lease updates are performed only by the `etcd` service account?

Reply:
Sounds like a good idea. That way we disallow changes to the lease object from human operators, and only allow etcdbr to do it.